### PR TITLE
New ergnomic client

### PR DESCRIPTION
**Pull request by johndagostino**
New AP Authentication Mechanism

### CODE REVIEW

File: src/resources/batch_transctions.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/* eslint-disable camelcase */
```
Comment (rrsmith) [inline src/resources/batch_transctions.ts]: typo in file name^

File: src/middleware.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    const canTry =
+      !config.expiredTokenRetry || config.expiredTokenRetry < retries;
+
+    if (!accessDenied && status === 401 && canTry) {
+      config.expiredTokenRetry += 1;
+
+      try {
+        await client.refresh();
+        config.headers.Authorization = `Bearer ${client?.token?.access_token}`;
+        return client.instance(config);
```
Comment (rrsmith) [inline src/middleware.ts:35]: Does this actually retry the original failed request? e.g. make a call to refund -> rejected due to token expiry -> refresh token -> ??

File: src/middleware.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import { Client } from './client';
+
+export function expiredToken(client: Client, retries = 5) {
+  return client.instance.interceptors.response.use(undefined, async error => {
```
Comment (rrsmith) [inline src/middleware.ts:4]: Trusting that axios knows to ignore `undefined` here (just wasn't in their limited set of examples, but then neither was the error handling function 🔥

File: src/middleware.ts (hunk 2 above)
Comment (rrsmith) [reply src/middleware.ts:35]: Answering my own question, yes, this should since config is the original request data

File: src/client.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    if (!this.token || !this.refreshedAt) {
+      await this.refresh();
+      return;
+    }
+
+    const expires = this?.token?.expires_in ?? 0;
+    // expires in the next 10 minutes
+    const ten = 1000 * 10 * 60;
```
Comment (rrsmith) [inline src/client.ts]: Any thoughts on using some jitter? The AP docs don't state a token limit, just a rate limit on requesting them.

File: src/middleware.ts (hunk 2 above)
Comment (johndagostino) [reply src/middleware.ts:35]: yup, will refresh and attempt the request again

File: src/middleware.ts (hunk 3 above)
Comment (johndagostino) [reply src/middleware.ts:4]: @rrsmith i trust the types :)

File: src/client.ts (hunk 4 above)
Comment (johndagostino) [reply src/client.ts]: @rrsmith yup good idea

**Review by rrsmith — APPROVED** [review]
Overall LGTM 👍

File: .eslintrc [hunk 5]
```diff
@@ -1,91 +1,5 @@
...
-        "**/*.ts"
-      ],
-      "rules": {
-        "no-undef": "warn",
-        "no-unused-vars": "off",
-        "import/no-unresolved": "error",
-        "import/no-extraneous-dependencies": "off",
-        "max-classes-per-file": "off"
-      }
-    }
```
Comment (ish-bindra) [inline .eslintrc:89]: 🎊

File: src/client.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
+/**
+ * @title Assembly API
+ * @version 2.0
+ * @baseUrl https://virtserver.swaggerhub.com/AssemblyPlatforms/assembly-api/2.0
```
Comment (ish-bindra) [inline src/client.ts:4]: The link isn't working.
Should be `https://api.swaggerhub.com/apis/AssemblyPlatforms/assembly-api/2.0/swagger.yaml`?

File: src/client.ts [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    });
+
+    this.authClient = new AuthClient({
+      baseURL: authBaseURL,
+      clientId,
+      clientSecret,
+      scope,
+    });
+
+    expiredToken(this, 3);
```
Comment (ish-bindra) [inline src/client.ts]: Nit: can make it a variable.

File: src/client.ts [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  }
+
+  getHeaders(secure = true) {
+    const headers = {
+      'User-Agent': `Assembly Payments Node 1.0`,
```
Comment (ish-bindra) [inline src/client.ts]: Version from package json?

File: src/resources/bank_accounts.ts [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+   * @request GET:/bank_accounts/{id}/users
+   * @secure
+   */
+  showBankAccountUser: (id: string) =>
+    client.request<SingleUser>({
+      url: `/bank_accounts/${id}/users`,
+      method: 'GET',
+      secure: true,
+    }),
+});
```
Comment (ish-bindra) [inline src/resources/bank_accounts.ts:108]: 😍

**Pull request merged**